### ISSUE TITLE
daylight saving time problem between 2 and 3 o'clock

### ISSUE THREAD

**Issue by jolichter** [modxcms/revolution#13397]
### Summary
TV Date Output: Instead 2 o'clock, I get 3 o'clock (only if summer time change!)

### Step to reproduce
create a TV Date and the Output Format is: %Y-%m-%d %R
`[[*EventDateStart]]`
- if summer time change (e.g. yesterday 2017-03-26)
- no matter what date I use, e.g.: 2017-03-26, 2017-04-12 ect.

### Observed behavior
- after saving the date between 2 and 3 o'clock and refresh the manager always 3 o'clock is saved
- all other hours working fine and I get the correct time.

### Expected behavior
Is this issue a smal MODX bug or a problem that Unixtime does not know daylight saving?

### Environment
MODX 2.5.5, php 5.6.23, apache 2.2.22, mysql 5.6.32

Link: [forums.modx.com](https://forums.modx.com/thread/101952/tv-date-output-instead-2-o-clock-i-get-3-o-clock#dis-post-549706)

**Comment by OptimusCrime** [comment]
Isn't this a PHP problem? MODX does not have any notion of timezones or daylight saving as far as I know. Check your server settings.

**Comment by jolichter** [comment]
Sorry, I try it more precise:

It is strange. The whole day when we switch to summer time:
MODX add an additional hour if I save a date between 2am and 3am for every date after daylight saving.

After the daylight saving (e.g. today) it is working normal.

**Comment by mrhaw** [comment]
What timezone are your PHP info displaying? Are you located in the same zone?
https://joshtronic.com/2013/11/04/handle-daylight-savings-time/

**Comment by jolichter** [comment]
Hi, my php timezone is Europe/Berlin and I'm located in this zone.
btw: MODX System Info, Server Time and Local Time is the same (Server Offset: 0)

Thank you all for your hints. I think this problem will occur only once a year and the next changing of Summer Time I will go in vacation ;-)

Please close this issue #13397. I think a discussion in the [forum](https://forums.modx.com/thread/101952/tv-date-output-instead-2-o-clock-i-get-3-o-clock#dis-post-549717) is sufficient for this?

**Comment by OptimusCrime** [comment]
If it turns out to be a MODX problem then we can reopen this and look into it. However, I am pretty sure that this is an issue with PHP and how it is set to handle daylight saving, because (as I already mentioned), MODX does not do anything special to handle daylight saving.

Because of that I think it makes most sense to take this in the forums.

Good luck :)

**Comment by Jako** [comment]
@jolichter What you could investigate: Which value is saved in the database? Is it saved right or already modified with the DST difference? It should be saved as date value in Revolution, so you could see the value directly. If it is saved right, we could look further in the MODX code. Otherwise it could be caused by ExtJS/Javascript date/time transformations.

**Comment by jolichter** [comment]
I figured out that problem is only if Winter Time switch to the Summer Time.
To be sure I installed a fresh MODX 2.5.6 (php 7.0.15) on my home pc without any plugins or extras. Here my steps to reproduce:

Today it is Sommertime (MESZ) 2017-03-28 about 23:15 o'clock
- create a TV Date [[*EventDateStart]] and the Output Format is: %Y-%m-%d %R
- create a new document

**Test 1:**
- set a future date/time in this TV: 2017-03-29 **2:00**
-> I get: 2017-03-29 **2:00**

**Test 2:**
- set a past date/time in this TV: 2017-03-25 **2:00**
-> I get: 2017-03-25 **2:00**

**Test 3:**
- set a past date/time (DST) in this TV: 2017-03-26 **2:00**
-> I get: 2017-03-26 **3:00**

This is correct: there is no such time between 2:00 and 2:59
_____________________________________________________

After that, I switched back the Linux system time to the 2017-03-26 10:00 (simulate the DST)

**Test 4:**
set a future date/time in this TV: 2017-03-29 **2:00**
-> I get: 2017-03-29 **3:00**

**Test 5:**
- set a past date/time in this TV: 2017-03-25 **2:00**
-> I get: 2017-03-25 **3:00**

**Test 6:**
- set a date/time in this TV: 2017-03-26 **2:00**
-> I get: 2017-03-26 **3:00**

MODX add an additional hour if I save a date between 2am and 3am for every date after daylight saving. I checked the value in DB live (modx_site_tmplvar_contentvalues) and MODX saved 3am instead of 2am.

**For my understanding ist Test 4 and Test 5 wrong. I hope this isn't a misunderstanding on my part?**

**Comment by mrhaw** [comment]
> After that, I switched back the Linux system time to the 2017-03-26 10:00 (simulate the DST)

Can you confirm you see the same time in MODX by running this snippet: 

`return date("D M j G:i:s T Y");`

**Comment by Jako** [comment]
If it is saved wrong, it could be caused by ExtJS/JavaScript. Could you investigate the posted values of the browser too.

**Comment by jolichter** [comment]
@mrhaw: yes I checked it with a snippet [[!today]]:
```
<?php
$today = date("Y-m-d H:i");
return $today;
```

```
my Document:
<p>MODX Revolution Version: [[++settings_version]]</p>
<p>Now is: [[!today]]</p>
<p>Test TV: [[*EventDateStart]]</p>
```

@Jako: I'll check it tomorrow. Do you use the "Firefox Developer Tools" for this?

**Comment by Jako** [comment]
I’m using Chrome Web Developer for this. But Firefox Developer Tools should be quite the same.

**Comment by mrhaw** [comment]
Posting this for clues
> Server Time here means what time PHP thinks it is, i.e. with the time() function. MODX stores and compares Resource-related dates and times as Unix timestamps, relative to Server Time. The server_offset_time is used to calculate Local Time only in the view.

https://docs.modx.com/revolution/2.x/administering-your-site/settings/system-settings/server_offset_time

**Comment by jolichter** [comment]
Hi, as far as I can see (I am not familiar with JS), do JS posted this wrong value?
I set 2am and JS post 3am (Test 5):
```
-----------------------------3561557801269986482670581215
Content-Disposition: form-data; name="tvs"

1
-----------------------------3561557801269986482670581215
Content-Disposition: form-data; name="tv1"

2017-03-25 03:00:00
```

[Here is a snapshot of it (Firefox-Developer)](http://spielwiese.jolichter.de/MODX_Firefox-Developer_DST-difference.png)
[Here is a snapshot of it (frontpage)](http://spielwiese.jolichter.de/MODX_Firefox-Developer_DST-difference_fontpage.png)

Yes it is really weird, but I can live with it ;-)

**Comment by jolichter** [comment]
PS: I am not sure, this parameter post in my snapshot is not from JS?

**Comment by OptimusCrime** [comment]
Could this be the cause of the problem: https://github.com/modxcms/revolution/blob/28fbe4400c117a06dcd1257cdaef8dab41a1a2bc/manager/assets/modext/widgets/resource/modx.panel.resource.js#L703 ?

**Comment by Jako** [comment]
It's a TV value and not publishedon. 

Since the value is posted wrong it looks like a javascript/browser bug.

**Comment by OptimusCrime** [comment]
You're right. It corresponds to this file: https://github.com/modxcms/revolution/blob/fbf126690f6ce903faeaeb9f1d958211fc2203c9/manager/templates/default/element/tv/renders/input/date.tpl

I'm pretty sure this is something ExtJS is messing up, as I belive the browser itself should not alter the actual value of a input field upon submit. I've never heard about this happening anywhere else. Thus, I find it most likely it is due to some ExtJS weirdness.

**Comment by mrhaw** [comment]
> Note: In countries which experience Daylight Saving Time (i.e. DST), the h, i, s and ms properties may coincide with the exact time in which DST takes effect. It is the responsiblity of the developer to account for this.
https://docs.sencha.com/extjs/3.4.0/?mobile=/api/Date

**Comment by Jako** [comment]
So it is a ExtJS bug that has to be solved in MODX somehow.

Just a few questions about this: When does the issue occur? Only between 2 o'clock and 3 o'clock on the day of DST change?

**Comment by jolichter** [comment]
No, on the whole day of DST change. The next day it works fine.

**Comment by Rajbir233** [comment]
Hi 
Did this issue got resolved? I am also facing similar kind of issue with Ext Js. Does anyone has any resolution to it?

**Comment by alroniks** [comment]
Nope, and I am not sure that we have to do that. If it ExtJS bug, it is still a rare case that happens on a specific date once a year. It is too expensive to spend time in case lack of contributors.

**Comment by jolichter** [comment]
Yes, that's weird and it only happens once a year... maybe this is solved in modx3?

**Comment by alroniks** [comment]
In MODX 3 still old ExtJS, so the issue seems will be left there as well.
